File: Patika.dev Coderbyte Assessments (Weekly)/Week 8/Java/BlackjackHighest.java
import java.io.*;
import java.util.*;

class Main {

    public static void main(String[] args) {
        // keep this function call here
        Scanner s = new Scanner(System.in);
        System.out.print(BlackjackHighest(s.nextLine()));
    }

    public static String BlackjackHighest(String[] strArr) {
        // code goes here
        Card[] cards = new Card[strArr.length];
        for (int i = 0; i < strArr.length; i++) {
            cards[i] = new Card(strArr[i]);
        }

        Arrays.sort(cards, new Comparator<Card>() {
            public int compare(Card card, Card card2) {
                if (card.rating > card2.rating) return 1;
                else if (card.rating < card2.rating) return -1;
                return 0;
            }
        });

        int total = 0;
        boolean hasAce = false, aceUsed = false;
        for (Card card : cards) {
            if (card.rating == 14) {
                hasAce = true;
            }
            if (total + card.topVal > 21) {
                total += card.lowVal;
            } else {
                if (card.rating == 14) {
                    aceUsed = true;
                }
                total += card.topVal;
            }
        }

        Card highest = cards[cards.length - 1];
        if (hasAce && !aceUsed) {
            highest = cards[cards.length - 2];
        }
        if (total < 21) {
            return "below " + highest.repr;
        } else if (total == 21) {
            return "blackjack " + highest.repr;
        }
        return "above " + highest.repr;
    }

    static class Card {
        int lowVal = 0, topVal = 0, rating = 0;
        String repr;

        Card(String card) {
            card = card.toLowerCase();
            repr = card;

            if (card.equals("two")) {
                lowVal = 2;
                topVal = 2;
                rating = 2;
            } else if (card.equals("three")) {
                lowVal = 3;
                topVal = 3;
                rating = 3;
            } else if (card.equals("four")) {
                lowVal = 4;
                topVal = 4;
                rating = 4;
            } else if (card.equals("five")) {
                lowVal = 5;
                topVal = 5;
                rating = 5;
            } else if (card.equals("six")) {
                lowVal = 6;
                topVal = 6;
                rating = 6;
            } else if (card.equals("seven")) {
                lowVal = 7;
                topVal = 7;
                rating = 7;
            } else if (card.equals("eight")) {
                lowVal = 8;
                topVal = 8;
                rating = 8;
            } else if (card.equals("nine")) {
                lowVal = 9;
                topVal = 9;
                rating = 9;
            } else if (card.equals("ten")) {
                lowVal = 10;
                topVal = 10;
                rating = 10;
            } else if (card.equals("jack")) {
                lowVal = 10;
                topVal = 10;
                rating = 11;
            } else if (card.equals("queen")) {
                lowVal = 10;
                topVal = 10;
                rating = 12;
            } else if (card.equals("king")) {
                lowVal = 10;
                topVal = 10;
                rating = 13;
            } else if (card.equals("ace")) {
                lowVal = 1;
                topVal = 11;
                rating = 14;
            }
        }
    }

}

File: Hard/Blackjack Highest.py
def BlackjackHighest(strArr):
    # code goes here
    values_blackjack = {"two": 2, "three": 3, "four": 4, "five": 5, "six": 6,
                        "seven": 7, "eight": 8, "nine": 9, "ten": 10,
                        "jack": 10, "queen": 10, "king": 10, "ace": 11}

    positioning_card = {"two": 2, "three": 3, "four": 4, "five": 5, "six": 6,
                        "seven": 7, "eight": 8, "nine": 9, "ten": 10,
                        "jack": 11, "queen": 12, "king": 13, "ace": 1}

    sum_val = 0
    highest_card_val = 0
    highest_card = ""

    for card in strArr:
        sum_val += values_blackjack[card]

        if positioning_card[card] > highest_card_val:
            highest_card_val = positioning_card[card]
            highest_card = card

    as_is_one = False
    if "ace" in strArr:
        if sum_val > 21:
            as_is_one = True
        else:
            highest_card = "ace"

    if as_is_one:
        sum_val -= 10

    if sum_val > 21:
        return "above {}".format(highest_card)
    elif sum_val == 21:
        return "blackjack {}".format(highest_card)
    else:
        return "below {}".format(highest_card)


# keep this function call here
print(BlackjackHighest(input()))
